File: programmers/lv1/크레인 인형 뽑기/크레인 인형 뽑기.py
# https://school.programmers.co.kr/learn/courses/30/lessons/64061


basket = []

def checkRemoveBall():
  result = 0
  temp = False
  while True:
    for i in range(len(basket) - 1):
      if basket[i] == basket[i+1]:  
        del basket[i:i + 2]
        result += 1
        temp = True
        break
    if not temp:
      break
    else:
      temp = False
      continue
  return result * 2


def solution(board, moves):
  n = len(board)
  result = 0
  for move in moves:
    for i in range(n):
      if board[i][move - 1] != 0:
        basket.append(board[i][move -1])
        board[i][move-1] = 0
        result+= checkRemoveBall()
        break
  return result




# 좋은 풀이
def solution(board, moves):
    stacklist = []
    answer = 0

    for i in moves:
        for j in range(len(board)):
            if board[j][i-1] != 0:
                stacklist.append(board[j][i-1])
                board[j][i-1] = 0

                if len(stacklist) > 1:
                    if stacklist[-1] == stacklist[-2]:
                        stacklist.pop(-1)
                        stacklist.pop(-1)
                        answer += 2     
                break

    return answer

File: programmers/lv1/크레인 인형 뽑기/크레인 인형 뽑기.js
// https://school.programmers.co.kr/learn/courses/30/lessons/64061

function solution(board, moves) {
  let result = 0;
  let stacklist = [];

  for (const move of moves) {
    for (let i = 0; i < board.length; i++) {
      if (board[i][move - 1] !== 0) {
        stacklist.push(board[i][move - 1]);
        board[i][move - 1] = 0;

        if (stacklist.length > 1) {
          const [x] = stacklist.slice(-1);
          const [y] = stacklist.slice(-2, -1);
          if (x === y) {
            stacklist.pop();
            stacklist.pop();

            result += 2;
          }
        }

        break;
      }
    }
  }

  return result;
}
